tools/CoqMakefile.in

```diff
@@ -709,9 +710,14 @@ $(addsuffix .d,$(MLPACKFILES)): %.mlpack.d: %.mlpack
 	$(SHOW)'COQDEP $<'
 	$(HIDE)$(COQDEP) $(OCAMLLIBS) -c "$<" $(redir_if_ok)
 
-$(addsuffix .d,$(VFILES)): %.v.d: %.v
```

Comment:
> otherwise, you'll get in infinite loops, I think.

Can you explain how that could happen?

Reply:
You could try:
```
$(addsuffix .d,$(VFILES)): vdfiles
	$(SHOW)'COQDEP $(VFILES)'
	$(HIDE)$(COQDEP) $(COQLIBS) -dyndep var -c -with-file-output $(VFILES) $(if_ok)
```

Reply:
> Can you explain how that could happen?

`vdfiles` is declared `PHONY`, which means that it is considered outdated at the start of every run of `make`.  Because each `.v.d` is `include`d in the `Makefile`, `make` will check to see if the `.v.d` files are out of date.  Because they depend on an outdated target (`vdfiles`), they are, so they get remade.  When an included makefile is remade, I believe `make` restarts its processing of the entire file, in particular reconsidering all `PHONY` targets as outdated.  Then we are back to the starting condition, and we have an infinite loop.

> You could try:

No.  Then `make -jN` will invoke that rule `N` times in parallel, which is bad.

I think this will avoid the infinite loops, though, without needing to modify coqdep:
```makefile
$(addsuffix .d,$(VFILES)): $(VFILES) | vdfiles
```
As I understand it, order-only prerequisites are explicitly for this sort of case (where it's still a prerequisite, but its "outdated" status doesn't transfer to the parent rule).

Reply:
These rules:
```
vdfiles: $(VFILES)
	$(SHOW)'COQDEP $^'
        $(HIDE)$(COQDEP) $(COQLIBS) -dyndep var -c -with-file-output $^

$(addsuffix .d,$(VFILES)): vdfiles
```
will call `coqdep` just once if you ask to build any number of `.v.d` files. But you get a "nothing to be done" message for each of them, which is annoying.

If you just run `make`, though, `make` gets called recursively, and `coqdep` gets called each time.

Reply:
With coq_makefile we have rules together for vo and glob, but if I do `make -j 2 foo.vo foo.glob` it doesn't run parallel coqc, why is that?
``` make
$(VOFILES): %.vo: %.v
	$(SHOW)COQC $<
	$(HIDE)$(TIMER) $(COQC) $(COQDEBUG) $(TIMING_ARG) $(COQFLAGS) $< $(TIMING_EXTRA)

# FIXME ?merge with .vo / .vio ?
$(GLOBFILES): %.glob: %.v
	$(TIMER) $(COQC) $(COQDEBUG) $(COQFLAGS) $<
```

Reply:
Same if you add the order-only prereqs.

Reply:
> Even without -jN, with that rule, if you try to build a particular .v.d file, coqdep gets run for every .v file.

I don't see that behavior.  You see two invocations of `coqdep` with this makefile?
```makefile
VFILES := foo.v bar.v

.PHONY: all
all: $(VFILES:.v=.vo)
        echo 1

$(VFILES:.v=.vo) : %.vo : %.v
        cp $< $@
        sleep 1

.PHONY: vdfiles
vdfiles:
        touch $(addsuffix .d,$(VFILES))

$(addsuffix .d,$(VFILES)): $(VFILES) | vdfiles

clean::
        rm -f $(VFILES:.v=.vo)
        rm -f $(addsuffix .d,$(VFILES))

-include $(addsuffix .d,$(VFILES))
```

Reply:
> With coq_makefile we have rules together for vo and glob, but if I do make -j 2 foo.vo foo.glob it doesn't run parallel coqc, why is that?

@SkySkimmer It does:
```
$ cat foo.v
Require Import Coq.ZArith.ZArith.

Time Definition foo0 := Eval vm_compute in Z.div_eucl.
$ cat _CoqProject
foo.v
$ coq_makefile -f _CoqProject -o Makefile3 && make -f Makefile3 clean && make -f Makefile3 -j2 foo.vo foo.glob
CLEAN
COQDEP foo.v
"coqc"  -q  -I .  -R . Top  foo.v
COQC foo.v
Finished transaction in 0.11 secs (0.095u,0.012s) (successful)
Finished transaction in 0.11 secs (0.092u,0.02s) (successful)
```